### PR TITLE
Thermalize particle momenta

### LINKED ISSUE #1472: `thermalize_particle_momenta` assigns velocities and angular momentum to constituent particles.
### Description

`thermalize_particle_momenta`, when given the `All` filter, will assign non-zero values to the velocities and angular momenta of rigid body constituent particles. `hoomd.md.constrain.Rigid` does not overwrite these values, though perhaps it should (see #278). Regardless, to behave consistently with `ThermodynamicQuantities`, `thermalize_particle_momenta` should assign 0's to constituent particle velocities and angular momenta:

https://github.com/glotzerlab/hoomd-blue/blob/925a73946736b4e77d31ba95c182b676798599bf/hoomd/ParticleGroup.cc#L824-L863

### Script

```python
import hoomd

snapshot = hoomd.Snapshot()
snapshot.configuration.box = (10, 10, 10, 0, 0, 0)
snapshot.particles.N = 4
snapshot.particles.types = ['A']
snapshot.particles.body[:] = [0, 0, 2, 2]
snapshot.particles.moment_inertia[:] = [[1,1,1]] * 4

sim = hoomd.Simulation(device=hoomd.device.CPU())
sim.create_state_from_snapshot(snapshot)

sim.state.thermalize_particle_momenta(filter=hoomd.filter.All(), kT=1.0)
thermalized_snapshot = sim.state.get_snapshot()
print(thermalized_snapshot.particles.velocity)
print(thermalized_snapshot.particles.angmom)
```

### Input files

_No response_

### Output

```shell
[[-0.66990062  0.13619617  0.63650922]
 [-1.00984745 -0.81729512  1.43739685]
 [ 0.77239644  0.5919247  -0.56883137]
 [ 0.90735163  0.08917425 -1.5050747 ]]
[[ 0.         -0.80604741  0.73100313  1.6675459 ]
 [ 0.         -1.39461841 -3.73358298 -1.2261137 ]
 [ 0.          2.98269913 -0.46326691  1.24673851]
 [ 0.          2.23798918  0.51482892  2.8423569 ]]
```

### Expected output

I expected 0 velocity and angular momentum for particles 1 and 3.

### Platform

CPU, GPU, Linux, macOS

### Installation method

Compiled from source

### HOOMD-blue version

3.8.1

### Python version

3.10.6

### PR DESCRIPTION
**Pull request by melodyyzh**
## Description
When the `All` filter is used on `thermalize_particle_momenta`, it assigns non-zero values to the velocities and angular momenta of rigid body constituent particles. These non-zero values are not be overwritten by `hoomd.md.constrain.Rigid`. This PR modified`thermalize_particle_momenta` to assign zero constituent particle velocities and angular momenta. 

## Motivation and context

Resolves #1472 

## How has this been tested?
Added a test function in `test_state.py` 

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by melodyyzh** [comment]
I have a couple of questions here: 
(1) in `ParticleGoup.cc`, when we loop around all the particles in a group, I believe these particles include both center and constituent particles. So I added an if statement by comparing `h_body` and `h_tag` to see if the particle is center or constituent. I am not sure if this logic is correct, since this change breaks the original 2 tests 
(`test_thermalize_particle_velocity` and `test_thermalize_angular_momentum`). 

(2) in the new test (`test_zero_particle_velocity_angmom`), I am currently using `thermalized_snapshot.particles.velocity.all()` to compare the velocity and angmom to 0. The test failed because velocity and angmom are not zero for **all** the particles. Is there some way I can restrict the comparison of constituent particles? 

Thank!!

File: hoomd/ParticleGroup.cc [hunk 1]
```diff
@@ -710,12 +710,16 @@ void ParticleGroup::thermalizeParticleMomenta(Scalar kT, uint64_t timestep)
                                    access_mode::read);
 
     ArrayHandle<unsigned int> h_tag(m_pdata->getTags(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                     access_location::host,
+                                     access_mode::readwrite);
```
Comment (tommy-waltmann) [inline hoomd/ParticleGroup.cc]:
```suggestion
                                     access_mode::read);
```
We're not going to modify the body tag data in this method, so just access it for reading.

File: hoomd/ParticleGroup.cc [hunk 2]
```diff
@@ -726,30 +730,42 @@ void ParticleGroup::thermalizeParticleMomenta(Scalar kT, uint64_t timestep)
                                                m_sysdef->getSeed()),
                                    hoomd::Counter(ptag));
 
-        // Generate a random velocity
+        // Generate zero velocities and momenta for constituent particles
```
Comment (tommy-waltmann) [inline hoomd/ParticleGroup.cc]:
```suggestion
        // Generate a random velocity, excluding constituent particles
```

File: hoomd/pytest/test_state.py [hunk 3]
```diff
@@ -223,6 +223,30 @@ def test_thermalize_angular_momentum(simulation_factory,
         assert K > expected_K * 3 / 4 and K < expected_K * 4 / 3
 
 
+def test_zero_particle_velocity_angmom():
+    snapshot = hoomd.Snapshot()
+    snapshot.configuration.box = (10, 10, 10, 0, 0, 0)
+    snapshot.particles.N = 4
+    snapshot.particles.types = ['A']
+    snapshot.particles.body[:] = [0, 0, 2, 2]
+    snapshot.particles.moment_inertia[:] = [[1, 1, 1]] * 4
```
Comment (tommy-waltmann) [inline hoomd/pytest/test_state.py]: Remember to check if the process is on rank 0 before updating the snapshot's particle data.

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
Great work on this @melodyyzh! I have one comment which will fix the failing unit tests with MPI, and two others which are mostly minor and won't affect unit testing results.

File: hoomd/ParticleGroup.cc [hunk 4]
```diff
@@ -726,30 +730,42 @@ void ParticleGroup::thermalizeParticleMomenta(Scalar kT, uint64_t timestep)
...
+        // Generate a random velocity, excluding constituent particles
         Scalar mass = h_vel.data[j].w;
         Scalar sigma = slow::sqrt(kT / mass);
         hoomd::NormalDistribution<Scalar> normal(sigma);
-        h_vel.data[j].x = normal(rng);
-        h_vel.data[j].y = normal(rng);
-        if (n_dimensions > 2)
-            h_vel.data[j].z = normal(rng);
+        // check if particles are constituent particles
+        if (h_tag.data[j] != h_body.data[j] && h_body.data[j] != static_cast<unsigned int>(-1))
```
Comment (joaander) [inline hoomd/ParticleGroup.cc]:
```suggestion
        if (h_tag.data[j] != h_body.data[j] && h_body.data[j] != NO_BODY)
```
Use defined constants in place of hard-coded values.

File: hoomd/ParticleGroup.cc [hunk 5]
```diff
@@ -726,30 +730,42 @@ void ParticleGroup::thermalizeParticleMomenta(Scalar kT, uint64_t timestep)
...
         quat<Scalar> q(h_orientation.data[j]);
         vec3<Scalar> I(h_inertia.data[j]);
 
-        if (I.x > 0)
-            p_vec.x = hoomd::NormalDistribution<Scalar>(slow::sqrt(kT * I.x))(rng);
-        if (I.y > 0)
-            p_vec.y = hoomd::NormalDistribution<Scalar>(slow::sqrt(kT * I.y))(rng);
-        if (I.z > 0)
-            p_vec.z = hoomd::NormalDistribution<Scalar>(slow::sqrt(kT * I.z))(rng);
+        if (h_tag.data[j] == h_body.data[j] || h_body.data[j] == static_cast<unsigned int>(-1))
```
Comment (joaander) [inline hoomd/ParticleGroup.cc]:
```suggestion
        if (h_tag.data[j] == h_body.data[j] || h_body.data[j] == NO_BODY)
```

File: hoomd/ParticleGroup.cc [hunk 6]
```diff
@@ -777,12 +793,15 @@ void ParticleGroup::thermalizeParticleMomenta(Scalar kT, uint64_t timestep)
         {
         unsigned int j = this->getMemberIndex(group_idx);
         Scalar mass = h_vel.data[j].w;
-        h_vel.data[j].x = h_vel.data[j].x - com_momentum.x / mass;
-        h_vel.data[j].y = h_vel.data[j].y - com_momentum.y / mass;
-        if (n_dimensions > 2)
-            h_vel.data[j].z = h_vel.data[j].z - com_momentum.z / mass;
-        else
-            h_vel.data[j].z = 0; // For 2D systems
+        if (h_tag.data[j] == h_body.data[j] || h_body.data[j] == static_cast<unsigned int>(-1))
```
Comment (joaander) [inline hoomd/ParticleGroup.cc]:
```suggestion
        if (h_tag.data[j] == h_body.data[j] || h_body.data[j] == NO_BODY)
```

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! I have suggestions to improve this.

**Review by tommy-waltmann — APPROVED** [review]
Once Josh's comments are addressed, this will be ready to go.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**